peers/prefer.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package peers
+
+import "github.com/uber/tchannel-go"
+
+type hrwScoreCalc struct {
+	clientID uint32
+}
+
+// NewHRWScorer returns a ScoreCalculator based on Rendezvouz or Highest Random Weight
+// hashing.
+// It is useful for distributing load in peer-to-peer situations where we have
+// many clients picking from a set of servers with "sticky" semantics that will
+// spread load evenly as servers go down or new servers are added.
+// The clientID is used to score the servers, so each client should pass in
+// a unique client ID.
+func NewHRWScorer(clientID uint32) tchannel.ScoreCalculator {
+	return &hrwScoreCalc{clientID}
+}
+
+func (s *hrwScoreCalc) GetScore(p *tchannel.Peer) uint64 {
+	server := fnv32a(p.HostPort())
+
+	// These constants are taken from W_rand2 in the Rendezvouz paper:
+	// http://www.eecs.umich.edu/techreports/cse/96/CSE-TR-316-96.pdf
+	return uint64(110315245*((1103515245*s.clientID+12345)^server) + 12345)
```

Comment:
Good point, it also suggests that the `clientID` and `server` should be a `2^31` hash, not sure how important that is.

I experimented a little to see what kind of distributions we get from `TestHRWScorerDistribution`

Current implementation: `[38 59 50 58 46 54 46 48 37 50 43 56 61 50 66 42 50 50 65 31]`
Final result modded by 2^31: `[45 51 40 56 40 49 36 56 38 56 54 67 55 56 46 46 48 49 61 51]`
Mod clientID, server, and final result by 2^31: `[45 51 40 56 40 49 36 56 38 56 54 67 55 56 46 46 48 49 61 51]`

I'll mod everything to match the paper. There's probably a different set of constants we could use to avoid mod completely and use `uint32` everywhere.